File: src/main/cc/wfa/virtual_people/common/field_filter/partial_filter_test.cc
// Copyright 2021 The Cross-Media Measurement Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "gmock/gmock.h"
#include "google/protobuf/text_format.h"
#include "gtest/gtest.h"
#include "src/main/cc/wfa/virtual_people/common/field_filter/test/test.pb.h"
#include "src/main/proto/wfa/virtual_people/common/field_filter.pb.h"
#include "wfa/virtual_people/common/field_filter/field_filter.h"

namespace wfa_virtual_people {
namespace {

using ::wfa_virtual_people::test::TestProto;

TEST(PartialFilterTest, TestNoName) {
  FieldFilterProto field_filter_proto;
  ASSERT_TRUE(google::protobuf::TextFormat::ParseFromString(R"PROTO(
      op: PARTIAL
      sub_filters {
        name: "int32_value"
        op: EQUAL
        value: "1"
      }
      sub_filters {
        name: "int64_value"
        op: EQUAL
        value: "1"
      }
  )PROTO", &field_filter_proto));
  auto field_filter_or =
      FieldFilter::New(TestProto().GetDescriptor(), field_filter_proto);
  EXPECT_EQ(field_filter_or.status().code(),
            absl::StatusCode::kInvalidArgument);
}

TEST(PartialFilterTest, TestNoSubFilters) {
  FieldFilterProto field_filter_proto;
  ASSERT_TRUE(google::protobuf::TextFormat::ParseFromString(R"PROTO(
      name: "a.b"
      op: PARTIAL
  )PROTO", &field_filter_proto));
  auto field_filter_or =
      FieldFilter::New(TestProto().GetDescriptor(), field_filter_proto);
  EXPECT_EQ(field_filter_or.status().code(),
            absl::StatusCode::kInvalidArgument);
}

TEST(PartialFilterTest, TestMatch) {
  FieldFilterProto field_filter_proto;
  ASSERT_TRUE(google::protobuf::TextFormat::ParseFromString(R"PROTO(
      name: "a.b"
      op: PARTIAL
      sub_filters {
        name: "int32_value"
        op: EQUAL
        value: "1"
      }
      sub_filters {
        name: "int64_value"
        op: EQUAL
        value: "1"
      }
  )PROTO", &field_filter_proto));
  auto field_filter_or =
      FieldFilter::New(TestProto().GetDescriptor(), field_filter_proto);
  EXPECT_TRUE(field_filter_or.ok());
  std::unique_ptr<FieldFilter> field_filter =
      std::move(field_filter_or.value());

  TestProto test_proto;
  ASSERT_TRUE(google::protobuf::TextFormat::ParseFromString(R"PROTO(
      a {
        b {
          int32_value: 1
          int64_value: 1
        }
      }
  )PROTO", &test_proto));
  EXPECT_TRUE(field_filter->IsMatch(test_proto));
}

TEST(PartialFilterTest, TestNotMatch) {
  FieldFilterProto field_filter_proto;
  ASSERT_TRUE(google::protobuf::TextFormat::ParseFromString(R"PROTO(
      name: "a.b"
      op: PARTIAL
      sub_filters {
        name: "int32_value"
        op: EQUAL
        value: "1"
      }
      sub_filters {
        name: "int64_value"
        op: EQUAL
        value: "1"
      }
  )PROTO", &field_filter_proto));
  auto field_filter_or =
      FieldFilter::New(TestProto().GetDescriptor(), field_filter_proto);
  EXPECT_TRUE(field_filter_or.ok());
  std::unique_ptr<FieldFilter> field_filter =
      std::move(field_filter_or.value());

  TestProto test_proto;
  ASSERT_TRUE(google::protobuf::TextFormat::ParseFromString(R"PROTO(
      a {
        b {
          int32_value: 1
          int64_value: 2
        }
      }
  )PROTO", &test_proto));
  EXPECT_FALSE(field_filter->IsMatch(test_proto));
}

}  // namespace
}  // namespace wfa_virtual_people

File: src/test/kotlin/org/wfanet/virtualpeople/common/fieldfilter/PartialFilterTest.kt
// Copyright 2022 The Cross-Media Measurement Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package org.wfanet.virtualpeople.common.fieldfilter

import kotlin.test.assertFailsWith
import kotlin.test.assertFalse
import kotlin.test.assertTrue
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4
import org.wfanet.virtualpeople.common.FieldFilterProto.Op
import org.wfanet.virtualpeople.common.fieldFilterProto
import org.wfanet.virtualpeople.common.test.*

@RunWith(JUnit4::class)
class PartialFilterTest {

  @Test
  fun `no name should fail`() {
    val fieldFilter = fieldFilterProto {
      op = Op.PARTIAL
      subFilters.add(
        fieldFilterProto {
          name = "int32_value"
          op = Op.EQUAL
          value = "1"
        }
      )
    }
    val exception =
      assertFailsWith<IllegalStateException> {
        FieldFilter.create(TestProto.getDescriptor(), fieldFilter)
      }
    assertTrue(exception.message!!.contains("Name must be set"))
  }

  @Test
  fun `repeated field should fail`() {
    val fieldFilter = fieldFilterProto {
      name = "a.b"
      op = Op.PARTIAL
      subFilters.add(
        fieldFilterProto {
          name = "int32_values"
          op = Op.EQUAL
          value = "1"
        }
      )
    }
    val exception =
      assertFailsWith<IllegalStateException> {
        FieldFilter.create(TestProto.getDescriptor(), fieldFilter)
      }
    assertTrue(exception.message!!.contains("Repeated field is not allowed"))
  }

  @Test
  fun `repeated field in the path should fail`() {
    val fieldFilter = fieldFilterProto {
      name = "repeated_proto_a.b"
      op = Op.PARTIAL
      subFilters.add(
        fieldFilterProto {
          name = "int32_value"
          op = Op.EQUAL
          value = "1"
        }
      )
    }
    val exception =
      assertFailsWith<IllegalStateException> {
        FieldFilter.create(TestProto.getDescriptor(), fieldFilter)
      }
    assertTrue(exception.message!!.contains("Repeated field is not allowed"))
  }

  @Test
  fun `no sub filter should fail`() {
    val fieldFilter = fieldFilterProto {
      name = "a.b"
      op = Op.PARTIAL
    }
    val exception =
      assertFailsWith<IllegalStateException> {
        FieldFilter.create(TestProto.getDescriptor(), fieldFilter)
      }
    assertTrue(exception.message!!.contains("sub_filters must be set"))
  }

  @Test
  fun `all subfilters match should pass`() {
    val fieldFilter = fieldFilterProto {
      name = "a.b"
      op = Op.PARTIAL
      subFilters.add(
        fieldFilterProto {
          name = "int32_value"
          op = Op.EQUAL
          value = "1"
        }
      )
      subFilters.add(
        fieldFilterProto {
          name = "int64_value"
          op = Op.EQUAL
          value = "1"
        }
      )
    }
    val filter = FieldFilter.create(TestProto.getDescriptor(), fieldFilter)
    val testProto1 = testProto {
      a = testProtoA {
        b = testProtoB {
          int32Value = 1
          int64Value = 1
        }
      }
    }
    assertTrue(filter.matches(testProto1))
    assertTrue(filter.matches(testProto1.toBuilder()))
  }

  @Test
  fun `any subfilter mismatch should fail`() {
    val fieldFilter = fieldFilterProto {
      name = "a.b"
      op = Op.PARTIAL
      subFilters.add(
        fieldFilterProto {
          name = "int32_value"
          op = Op.EQUAL
          value = "1"
        }
      )
      subFilters.add(
        fieldFilterProto {
          name = "int64_value"
          op = Op.EQUAL
          value = "1"
        }
      )
    }
    val filter = FieldFilter.create(TestProto.getDescriptor(), fieldFilter)
    val testProto1 = testProto {
      a = testProtoA {
        b = testProtoB {
          int32Value = 1
          int64Value = 2
        }
      }
    }
    assertFalse(filter.matches(testProto1))
    assertFalse(filter.matches(testProto1.toBuilder()))
  }

  @Test
  fun `field not set should fail`() {
    val fieldFilter = fieldFilterProto {
      name = "a.b"
      op = Op.PARTIAL
      subFilters.add(
        fieldFilterProto {
          name = "int32_value"
          op = Op.EQUAL
          value = "0"
        }
      )
    }
    val filter = FieldFilter.create(TestProto.getDescriptor(), fieldFilter)
    val testProto1 = testProto {}
    assertFalse(filter.matches(testProto1))
    assertFalse(filter.matches(testProto1.toBuilder()))
  }
}
